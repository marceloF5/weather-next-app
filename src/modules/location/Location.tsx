import { useEffect, useRef, useState } from 'react'

import { Stack } from 'components/base'
import { Field, ListBox, ListItem, Text } from 'components/ui'
import { useWeather } from 'shared/context/weather'
import { TCityInfoMapped } from 'shared/types'

const Location = () => {
    const {
        listOfCitiesFound,
        handleListOfCities,
        handleCurrentWeatherByCoordinates,
        resetListOfCitiesFound
    } = useWeather()
    const [errorMsg, setErrorMsg] = useState<string | null>(null)
    const isResultListOpen = !!(listOfCitiesFound.length > 1)
    const searchByCityRef = useRef<HTMLInputElement>(null)
    const listOfCitiesFoundWithRef = useRef<HTMLLIElement[]>([])
    searchByCityRef.current?.focus()

    useEffect(() => {
        listOfCitiesFoundWithRef.current =
            listOfCitiesFoundWithRef.current.slice(0, listOfCitiesFound.length)
    }, [listOfCitiesFound.length])

    const handleKeyDown = (index: number) => {
        if (listOfCitiesFoundWithRef.current.length > 1) {
            const nextElement = listOfCitiesFoundWithRef.current[index]
                .nextElementSibling as HTMLLIElement
            const hasNextElement = !!nextElement

            hasNextElement
                ? nextElement.focus()
                : listOfCitiesFoundWithRef.current[0].focus()
        }
    }

    const handleFocusToInput = () => searchByCityRef.current?.focus()

    const handleFocusToList = () =>
        listOfCitiesFoundWithRef.current.length
            ? listOfCitiesFoundWithRef.current[0].focus()
            : setErrorMsg('Please, search by city, state or country')

    const handleResetListOfCities = () => {
        resetListOfCitiesFound()
        if (searchByCityRef.current) {
            searchByCityRef.current.value = ''
        }
    }

    const handleSelectCity = (cityinfo: TCityInfoMapped) => {
        handleCurrentWeatherByCoordinates(cityinfo)

        if (searchByCityRef.current) {
            searchByCityRef.current.value = ''
        }
    }

    const handleKeyUp = (index: number) => {
        if (listOfCitiesFoundWithRef.current.length > 1) {
            const previousElement = listOfCitiesFoundWithRef.current[index]
                .previousElementSibling as HTMLLIElement
            const hasPreviousElement = !!previousElement

            hasPreviousElement
                ? previousElement.focus()
                : listOfCitiesFoundWithRef.current[
                      listOfCitiesFoundWithRef.current.length - 1
                  ].focus()
        }
    }

    return (
        <Stack align="center" gap="2" position="relative">
            <Stack gap="1" direction="vertical">
                <Field
                    type="text"
                    name="searchByCity"
                    placeholder="search city..."
                    onChange={(ev) => {
                        setErrorMsg(null)
                        handleListOfCities(ev.target.value)
                    }}
                    onKeyDown={(ev) => {
                        if (ev.code === 'Enter') handleFocusToList()
                        if (ev.code === 'Escape') handleResetListOfCities()
                        if (ev.code === 'ArrowDown') handleFocusToList()
                    }}
                    ref={searchByCityRef}
                    tabIndex={0}
                    autoComplete="off"
                />
                {!!errorMsg && (
                    <Text size="1.5rem" weight="bold" hasError>
                        {errorMsg}
                    </Text>
                )}
            </Stack>

            <ListBox isOpen={isResultListOpen}>
                {listOfCitiesFound.map((cityinfo, i) => (
                    <ListItem
                        ref={(el) => {
                            el && (listOfCitiesFoundWithRef.current[i] = el)
                        }}
                        tabIndex={0}
                        key={cityinfo.latitude + cityinfo.longitude}
                        onKeyDown={(ev) => {
                            if (ev.code === 'Enter') handleSelectCity(cityinfo)
                            if (ev.code === 'Escape') handleFocusToInput()
                            if (ev.code === 'ArrowDown') handleKeyDown(i)
                            if (ev.code === 'ArrowUp') handleKeyUp(i)
                        }}
                        onClick={() => handleSelectCity(cityinfo)}
                    >
                        <p>
                            {cityinfo.city}
                            {cityinfo.state ? ', ' + cityinfo.state : ''}
                            {cityinfo.country ? ', ' + cityinfo.country : ''}
                        </p>
                    </ListItem>
                ))}
            </ListBox>
        </Stack>
    )
}

export default Location
