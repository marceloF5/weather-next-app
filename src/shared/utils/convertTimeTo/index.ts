function convertTimeToDateOfWeek(dayWeek: number) {
    return new Date(dayWeek * 1000).toLocaleDateString('en-US', {
        weekday: 'short'
    })
}

function convertTimeToHour(time: number) {
    return new Date(time * 1000).toLocaleTimeString('en-US', {
        hour: '2-digit',
        minute: '2-digit'
    })
}

export { convertTimeToDateOfWeek, convertTimeToHour }
