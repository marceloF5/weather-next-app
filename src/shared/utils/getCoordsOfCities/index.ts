import { TCityInfoMapped, TCityInfoResponse } from 'shared/types'

async function getCoordsOfCities(city: string): Promise<TCityInfoMapped[]> {
    const url = `${process.env.OPEN_WEATHER_API_URL_CITY}q=${city}&limit=10&appid=${process.env.OPEN_WEATHER_TOKEN}`
    const response = await fetch(url)
    const cityInfoResponse = (await response.json()) as TCityInfoResponse[]

    try {
        if (!cityInfoResponse.length) throw new Error('No city info found')

        const cityInfoMapped = cityInfoResponse.map((info) => ({
            city: info.name,
            state: info.state,
            country: info.country,
            latitude: info.lat,
            longitude: info.lon
        }))

        return cityInfoMapped
    } catch (_) {
        // This catch could be improved with real error message
        throw new Error('API unavailable')
    }
}

export default getCoordsOfCities
