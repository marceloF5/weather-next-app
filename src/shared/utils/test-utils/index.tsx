import { render, RenderOptions } from '@testing-library/react'
import { ReactElement } from 'react'

import { Container } from 'components/base'
import { WeatherProvider } from 'shared/context/weather'
import { ThemeProvider } from 'styled-components'
import GlobalStyles from 'styles/global'
import theme from 'styles/theme'

import {
    initial_city_info_mock,
    initial_list_of_cities_found_mock
} from './page-mock'

type CustomRenderProps = {
    initial_city_info?: string
    initial_list_of_cities_found?: string
} & Omit<RenderOptions, 'queries'>

const customRender = (
    ui: ReactElement,
    {
        initial_city_info = JSON.stringify(initial_city_info_mock),
        initial_list_of_cities_found = JSON.stringify(
            initial_list_of_cities_found_mock
        ),
        ...renderOptions
    }: CustomRenderProps = {}
) =>
    render(
        <ThemeProvider theme={theme}>
            <GlobalStyles />
            <WeatherProvider
                initial_city_info={initial_city_info}
                initial_list_of_cities_found={initial_list_of_cities_found}
            >
                <Container>{ui}</Container>
            </WeatherProvider>
        </ThemeProvider>,
        renderOptions
    )

export * from '@testing-library/react'
export { customRender }
