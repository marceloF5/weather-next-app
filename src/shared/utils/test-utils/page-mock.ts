export const initial_city_info_mock =
    '{lat:4.7429,lon:-74.123,city:Lisboa,country:CO}'
export const initial_list_of_cities_found_mock =
    '[{city:London,state:England,country:GB,latitude:51.5073219,longitude:-0.1276474,temp:19%C2%BA,id:51.3796745},{city:S%C3%A3o%20Paulo,state:S%C3%A3o%20Paulo,country:BR,latitude:-23.5506507,longitude:-46.6333824,temp:25%C2%BA,id:-70.1840331},{city:Lisbon,country:PT,latitude:38.7077507,longitude:-9.1365919,temp:25%C2%BA,id:29.5711588}]'
