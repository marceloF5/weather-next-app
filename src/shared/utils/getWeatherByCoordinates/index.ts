import {
    TCoordinates,
    TForecast,
    TWeatherMapped,
    TWeatherResponse
} from 'shared/types'

import { convertTimeToDateOfWeek, convertTimeToHour } from '../convertTimeTo'

async function getWeatherByCoordinates(
    lat?: number,
    lon?: number
): Promise<TWeatherMapped> {
    let url = ''

    if (!lat || !lon) {
        const userLocation = await new Promise<TCoordinates>(
            (resolve, reject) => {
                navigator.geolocation.getCurrentPosition(resolve, reject)
            }
        )
        const {
            coords: { latitude, longitude }
        } = userLocation
        url = `${process.env.OPEN_WEATHER_API_URL_COORD}lat=${latitude}&lon=${longitude}&exclude=hourly,minutely&units=metric&appid=${process.env.OPEN_WEATHER_TOKEN}`
    } else {
        url = `${process.env.OPEN_WEATHER_API_URL_COORD}lat=${lat}&lon=${lon}&exclude=hourly,minutely&units=metric&appid=${process.env.OPEN_WEATHER_TOKEN}`
    }

    const response = await fetch(url)
    const weatherInfo = (await response.json()) as TWeatherResponse

    try {
        if (!weatherInfo) throw new Error('No weather info found')

        const forecasts = weatherInfo.daily.slice(1, 7).map((item) => {
            return {
                dt: convertTimeToDateOfWeek(item.dt),
                temp: `${item.temp.day.toFixed()}ยบ`
            } as TForecast
        })

        return {
            timezone: weatherInfo.timezone,
            latitude: weatherInfo.lat,
            longitude: weatherInfo.lon,
            dt: convertTimeToDateOfWeek(weatherInfo.current.dt),
            temp: `${weatherInfo.current.temp.toFixed()}ยบ`,
            feelsLike: `${weatherInfo.current.feels_like.toFixed()}ยบ`,
            humidity: `${weatherInfo.current.humidity.toFixed()}%`,
            pressure: `${weatherInfo.current.pressure.toFixed()} MB`,
            sunrise: convertTimeToHour(weatherInfo.current.sunrise),
            sunset: convertTimeToHour(weatherInfo.current.sunset),
            windSpeed: `${weatherInfo.current.wind_speed.toFixed(2)} km/h`,
            forecasts
        }
    } catch (_) {
        // This catch could be improved with real error message
        throw new Error('API unavailable')
    }
}

export default getWeatherByCoordinates
