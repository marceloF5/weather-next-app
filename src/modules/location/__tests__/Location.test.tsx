import '@testing-library/jest-dom'

import { customRender } from 'shared/utils/test-utils'

import Location from '../Location'

describe('<Location />', () => {
    it('render a snapshot of the component', () => {
        const { container } = customRender(<Location />)

        expect(container).toMatchSnapshot()
    })
})
