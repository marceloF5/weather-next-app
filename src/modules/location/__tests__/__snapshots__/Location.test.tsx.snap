// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Location /> render a snapshot of the component 1`] = `
.c0 {
  height: 100vh;
  width: 100%;
  margin: auto;
  padding: 2.5rem;
  background-color: #000;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c1 {
  width: 100%;
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  gap: 2rem;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
}

.c2 {
  width: 100%;
  max-width: 50rem;
  background-color: #1a191c;
  padding: 1rem;
  border-radius: 0.5rem;
  border: solid 0.2rem #ccc;
  font-size: xx-large;
}

.c3 {
  width: 100%;
  max-width: 50rem;
  background-color: #1a191c;
  padding: 2rem 0;
  border-radius: 0.5rem;
  -webkit-transform: scaleY(0) translateY(0.05rem);
  -ms-transform: scaleY(0) translateY(0.05rem);
  transform: scaleY(0) translateY(0.05rem);
  -webkit-transform-origin: top;
  -ms-transform-origin: top;
  transform-origin: top;
  -webkit-transition: -webkit-transform 0.2s ease;
  -webkit-transition: transform 0.2s ease;
  transition: transform 0.2s ease;
  position: absolute;
  top: 7rem;
}

@media (max-width:450px) {
  .c0 {
    height: calc(100vh - calc(100vh - 100%));
  }
}

@media (min-width:768px) and (max-width:1170px) {
  .c0 {
    padding: 2.5 3rem;
  }
}

@media (min-width:1170px) {
  .c0 {
    padding: 5rem 10rem;
  }
}

@media (min-width:1440px) {
  .c0 {
    padding: 5rem 16rem;
  }
}

<div>
  <div
    class="c0"
  >
    <div
      class="c1"
    >
      <input
        class="c2"
        name="searchByCity"
        placeholder="search city..."
        type="text"
      />
      <ul
        class="c3"
      />
    </div>
  </div>
</div>
`;
